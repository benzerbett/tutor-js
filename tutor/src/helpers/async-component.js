import { asyncComponent as asyncLoader } from 'react-async-component';
import ErrorComponent from '../components/error-monitoring/async-load-error';
import LoadingComponent from '../components/loading-screen';

export function asyncComponent(resolve) {
  return asyncLoader({
    resolve,
    LoadingComponent,
    ErrorComponent,
  });
}

export function loadAsync(resolve) {
  // return a function so the router will only evaluate it when it's needed
  // in the future we can insert role dependant logic here
  return () => asyncComponent(resolve);
}
