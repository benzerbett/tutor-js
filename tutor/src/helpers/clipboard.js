import _ from 'underscore';

const SUPPORTED = typeof document.queryCommandSupported === 'function' ? document.queryCommandSupported('copy') : undefined;

export default {
  isSupported: _.constant(!!SUPPORTED),

  copy() {
    if (!SUPPORTED) { return; }

    try {
      return document.execCommand('copy');
    } catch (e) {
      return console.warn('clipboard copy failed', e);
    }
  },
};
