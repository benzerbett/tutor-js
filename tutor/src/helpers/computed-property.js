import { forIn } from 'lodash';
import { computed } from 'mobx';

export function addComputedProperty(obj, name, fn) {

  Object.defineProperty(obj, name, {
    configurable: true,
    get() {
      const getter = computed(fn, { context: this });
      Object.defineProperty(this, name, {
        configurable: false,
        get: () => getter.get(),
      });
      return getter.get();
    },
  });

}

export function extendHasMany(properties) {
  return (collection) => {
    forIn(properties, (fn, prop) => {
      addComputedProperty(collection, prop, fn);
    });
  };
}
