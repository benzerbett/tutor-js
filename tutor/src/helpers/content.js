import _ from 'underscore';


export default {
  // Convert each number to base 10 with it's position based on index.
  // If section is not present, 0 is set for it
  //   1 becomes 100, 1.1 becomes 101, 3.2 to 302, 3.2.1 -> 30201
  // Useful for sorting
  chapterSectionToNumber(chapter_section) {
    if (_.isString(chapter_section)) { chapter_section = chapter_section.split('.'); }
    if (chapter_section.length === 1) { chapter_section.push(0); } // add a section 0 if it has only a chapter
    let position = -1;
    return _.reduceRight(chapter_section, (memo, num) => memo + (num * Math.pow(100, (position += 1)))
      , 0);
  },
};
