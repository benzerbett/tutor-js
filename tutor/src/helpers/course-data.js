import Courses from '../models/courses-map';

import Router from '../helpers/router';
import TimeHelper from '../helpers/time';

export default {
  getCourseDataProps(courseId) {
    let dataProps;
    if (courseId == null) {
      ((((((({ courseId } = Router.currentParams())))))));
    }
    const course = Courses.get(courseId);
    return dataProps = {
      'data-title': course.name,
      'data-book-title': course.bookName || '',
      'data-appearance': course.appearance_code,
    };
  },

  getCourseBounds(courseId) {
    const course = Courses.get(courseId);

    const start = TimeHelper.getMomentPreserveDate(course.starts_at, TimeHelper.ISO_DATE_FORMAT);
    const end = TimeHelper.getMomentPreserveDate(course.ends_at, TimeHelper.ISO_DATE_FORMAT);

    return { start, end };
  },
};
