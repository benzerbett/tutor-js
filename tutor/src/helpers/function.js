export default {
  // coffeelint: disable=no_empty_functions
  empty() {},
};
// coffeelint: enable=no_empty_functions
