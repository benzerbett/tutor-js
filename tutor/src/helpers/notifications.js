import { NotificationActions } from 'shared';

export default {
  start(bootstrapData) {
    NotificationActions.startPolling();
    return (() => {
      const result = [];
      const object = bootstrapData.flash || {};
      for (let level in object) {
        const message = object[level];
        result.push(NotificationActions.display({ message, level }));
      }
      return result;
    })();
  },

  buildCallbackHandlers(comp) {
    // Require router here because requiring it in global scope
    // causes hot module reloading to fail in dev mode because this file is required outside the HMR scope
    const Router = require('./router');
    const { router } = comp.context;
    if (!router) {
      throw new Error('Component\'s context must have router present');
    }
    return {
      missing_student_id: {
        onAdd({ course }) {
          return router.history.push(
            Router.makePathname('changeStudentId', { courseId: course.id })
          );
        },
      },
    };
  },
};
