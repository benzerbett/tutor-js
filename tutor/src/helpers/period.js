import { filter, isEmpty, sortBy, toNumber, isNumber } from 'lodash';
import { addComputedProperty } from './computed-property';
import S from './string';

// Used to filter periods by helper methods
const isArchivedCheckFn = period => period.is_archived;
const isActiveCheckFn   = period => !period.is_archived;

const NumberLike = /[^0-9]+/;


const PeriodHelper = {
  getOrder(period) {
    return S.getNumberAndStringOrder(period.name);
  },

  sort(periods) {
    // expects either numbers, names with numbers or just names
    periods = sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const name = period.name.match(/[^0-9]+/ig);
        if (name) {
          return name;
        }
      }
    });
    return sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const number = period.name.match(/[0-9.-]+/g);
        if (number) {
          return parseFloat(number);
        }
      } else {
        return period.name;
      }
    });
  },

};


export default PeriodHelper;
