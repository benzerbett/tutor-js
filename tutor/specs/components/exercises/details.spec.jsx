import { React, SnapShot } from 'helpers';
import Details from '../../../src/components/exercises/details';
import FakeWindow from 'shared/specs/helpers/fake-window';
import Factory from '../../factories';

jest.mock('../../../../shared/src/components/html', () => ({ html }) =>
  html ? <div dangerouslySetInnerHTML={{ __html: html }} /> : null
);

describe('Exercise Details Component', function() {

  let exercises, props, book;

  beforeEach(() => {
    book = Factory.book();
    const pageIds = book.pages.byId.keys();
    exercises = Factory.exercisesMap({ book, pageIds, count: 4 });
    props = {
      book,
      exercises,
      pageIds,
      windowImpl:             new FakeWindow,
      selectedExercise:       exercises.array[1],
      onExerciseToggle:       jest.fn(),
      getExerciseIsSelected:  jest.fn().mockReturnValue(false),
      getExerciseActions:     jest.fn().mockReturnValue({}),
      onShowCardViewClick:    jest.fn(),
    };
  });


  it('renders and matches snapshot', () => {
    const component = SnapShot.create(<Details {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('can navigate back/forward', () => {
    const deets = mount(<Details {...props} />);
    expect(deets).toHaveRendered(`[data-exercise-id="${exercises.array[1].content.uid}"]`);
    deets.find('.paging-control.next').simulate('click');
    expect(deets).toHaveRendered(`[data-exercise-id="${exercises.array[2].content.uid}"]`);
    deets.find('.paging-control.prev').simulate('click');
    deets.find('.paging-control.prev').simulate('click');
    expect(deets).toHaveRendered(`[data-exercise-id="${exercises.array[0].content.uid}"]`);
    expect(deets).toHaveRendered('.paging-control.prev[disabled]');
  });

});
