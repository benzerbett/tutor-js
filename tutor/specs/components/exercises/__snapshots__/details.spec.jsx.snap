// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exercise Details Component renders and matches snapshot 1`] = `
<div
  className="exercise-details"
>
  <div
    className="controls"
  >
    <a
      className="show-cards"
      onClick={[Function]}
    >
      <i
        className="tutor-icon fa fa-th-large"
        role="presentation"
        type="th-large"
      />
       Back to Card View
    </a>
  </div>
  <div
    className="content"
  >
    <div
      className="tutor-paging-navigation"
    >
      <a
        aria-controls="paged-content"
        className="paging-control prev"
        disabled={false}
        href="#"
        onClick={[Function]}
        tabIndex={0}
        target="_blank"
        title="Go Back"
      >
        <div
          className="arrow-wrapper"
        >
          <svg
            className="icon arrow left"
            version="1.1"
            viewBox="0 0 370 735"
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            y="0px"
          >
            <path
              d="M369.225,658.604l0,76.396l-369.225,-369.225l0.3,-0.3l368.925,-365.475l0,77.168l-291.449,290.332l291.449,291.104Z"
            />
          </svg>
        </div>
      </a>
      <div
        aria-live="polite"
        className="paged-content"
        id="paged-content"
        role="region"
        tabIndex="0"
      >
        <div
          className="openstax-exercise-preview exercise-card actions-on-side panel panel-default"
          data-exercise-id="2@1"
          id={undefined}
          tabIndex={-1}
        >
          <div
            className="panel-body"
          >
            <div
              className="exercise-body"
            >
              <div
                className="openstax-question openstax-question-preview"
                data-question-number={undefined}
              >
                <div
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "How could physics be useful in weather prediction?",
                    }
                  }
                />
                <div
                  className="answers-table"
                >
                  <div
                    className="openstax-answer"
                  >
                    <section
                      className="answers-answer disabled"
                      role="region"
                    >
                      <label
                        className="answer-label"
                        htmlFor="2-option-0"
                        onKeyPress={[Function]}
                      >
                        <button
                          aria-label="Choice a:"
                          className="answer-letter"
                          disabled={true}
                          onClick={undefined}
                        >
                          a
                        </button>
                        <div
                          className="answer-answer"
                        >
                          <div
                            dangerouslySetInnerHTML={
                              Object {
                                "__html": "Physics helps in predicting dynamics and movement of weather phenomena.",
                              }
                            }
                          />
                        </div>
                      </label>
                    </section>
                  </div>
                  <div
                    className="openstax-answer"
                  >
                    <section
                      className="answers-answer disabled"
                      role="region"
                    >
                      <label
                        className="answer-label"
                        htmlFor="2-option-1"
                        onKeyPress={[Function]}
                      >
                        <button
                          aria-label="Choice b:"
                          className="answer-letter"
                          disabled={true}
                          onClick={undefined}
                        >
                          b
                        </button>
                        <div
                          className="answer-answer"
                        >
                          <div
                            dangerouslySetInnerHTML={
                              Object {
                                "__html": "Physics helps in predicting how burning fossil fuel releases pollutants.",
                              }
                            }
                          />
                        </div>
                      </label>
                    </section>
                  </div>
                  <div
                    className="openstax-answer"
                  >
                    <section
                      className="answers-answer disabled"
                      role="region"
                    >
                      <label
                        className="answer-label"
                        htmlFor="2-option-2"
                        onKeyPress={[Function]}
                      >
                        <button
                          aria-label="Choice c:"
                          className="answer-letter"
                          disabled={true}
                          onClick={undefined}
                        >
                          c
                        </button>
                        <div
                          className="answer-answer"
                        >
                          <div
                            dangerouslySetInnerHTML={
                              Object {
                                "__html": "Physics helps in predicting how the flowing water affects Earthâ€™s surface.",
                              }
                            }
                          />
                        </div>
                      </label>
                    </section>
                  </div>
                  <div
                    className="openstax-answer"
                  >
                    <section
                      className="answers-answer disabled"
                      role="region"
                    >
                      <label
                        className="answer-label"
                        htmlFor="2-option-3"
                        onKeyPress={[Function]}
                      >
                        <button
                          aria-label="Choice d:"
                          className="answer-letter"
                          disabled={true}
                          onClick={undefined}
                        >
                          d
                        </button>
                        <div
                          className="answer-answer"
                        >
                          <div
                            dangerouslySetInnerHTML={
                              Object {
                                "__html": "Physics helps in predicting the motion of tectonic plates.",
                              }
                            }
                          />
                        </div>
                      </label>
                    </section>
                  </div>
                  <div
                    className="openstax-answer"
                  >
                    <section
                      className="answers-answer disabled"
                      role="region"
                    >
                      <label
                        className="answer-label"
                        htmlFor="2-option-4"
                        onKeyPress={[Function]}
                      >
                        <button
                          aria-label="Choice e:"
                          className="answer-letter"
                          disabled={true}
                          onClick={undefined}
                        >
                          e
                        </button>
                        <div
                          className="answer-answer"
                        >
                          <div
                            dangerouslySetInnerHTML={
                              Object {
                                "__html": "Physics helps in predicting dynamics and movement of weather phenomena.",
                              }
                            }
                          />
                        </div>
                      </label>
                    </section>
                  </div>
                </div>
                <div
                  className="detailed-solution"
                >
                  <div
                    className="header"
                  >
                    Detailed solution
                  </div>
                  <div
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "four",
                      }
                    }
                  />
                </div>
              </div>
            </div>
            <div
              className="exercise-tags"
            >
              <span
                className="exercise-tag"
              >
                blooms:3
              </span>
              <span
                className="exercise-tag"
              >
                time:long
              </span>
              <span
                className="exercise-tag"
              >
                dok:3
              </span>
              <span
                className="exercise-tag"
              >
                lo:stax-phys:1-2-1
              </span>
              <span
                className="exercise-tag"
              >
                ID:2@1
              </span>
            </div>
          </div>
        </div>
      </div>
      <a
        aria-controls="paged-content"
        className="paging-control next"
        disabled={false}
        href="#"
        onClick={[Function]}
        tabIndex={0}
        title="Go Forward"
      >
        <div
          className="arrow-wrapper"
        >
          <svg
            className="icon arrow right"
            version="1.1"
            viewBox="0 0 370 735"
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            y="0px"
          >
            <path
              d="M0,658.604l0,76.396l369.225,-369.225l-0.3,-0.3l-368.925,-365.475l0,77.168l291.449,290.332l-291.449,291.104Z"
            />
          </svg>
        </div>
      </a>
    </div>
  </div>
</div>
`;
